staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/fieldmanager.go

```diff
@@ -282,3 +287,65 @@ func (f *FieldManager) stripFields(managed fieldpath.ManagedFields, manager stri
 
 	return managed
 }
+
+// TODO: Determine a better value for this constant.
+const maxUpdateManagers int = 10
+
+// capUpdateManagers merges a number of the oldest update entries into versioned buckets,
+// such that the number of entries from updates does not exceed 'maxUpdateManagers'
+func (f *FieldManager) capUpdateManagers(managed internal.Managed) (newManaged internal.Managed, err error) {
+	// Gather all entries from updates
+	updaters := []string{}
+	for manager, fields := range managed.Fields {
+		if fields.Applied() == false {
+			updaters = append(updaters, manager)
+		}
+	}
+	if len(updaters) <= maxUpdateManagers {
+		return managed, nil
+	}
+
+	// If we have more than the maximum, sort the update entries by time, oldest first.
+	sort.Slice(updaters, func(i, j int) bool {
+		iTime, jTime, nTime := managed.Times[updaters[i]], managed.Times[updaters[j]], &metav1.Time{Time: time.Time{}}
+		if iTime == nil {
+			iTime = nTime
+		}
+		if jTime == nil {
+			jTime = nTime
+		}
+		if !iTime.Equal(jTime) {
+			return iTime.Before(jTime)
+		}
+		return updaters[i] < updaters[j]
+	})
+
+	// Merge the oldest updaters with versioned bucket managers until the number of updaters is under the cap
+	for i := 0; i < len(updaters)-maxUpdateManagers; i++ {
+		manager := updaters[i]
+		vs := managed.Fields[manager]
+		time := managed.Times[manager]
+		delete(managed.Fields, manager)
+
+		// Create a new manager identifier for the versioned bucket entry.
+		// The version for this manager comes from the version of the update being merged into the bucket.
+		if manager, err = internal.BuildManagerIdentifier(&metav1.ManagedFieldsEntry{
+			Manager:    "manager-too-old-to-track",
+			Operation:  metav1.ManagedFieldsOperationUpdate,
+			APIVersion: string(vs.APIVersion()),
+		}); err != nil {
+			return managed, fmt.Errorf("failed to create bucket manager for version %v: %v", string(vs.APIVersion()), err)
+		}
+
+		// Merge the fieldets if the bucket already exists, otherwise just create a new entry.
+		// Use the time from the update being merged into the bucket, since it is more recent.
+		if previous, ok := managed.Fields[manager]; ok {
+			managed.Fields[manager] = fieldpath.NewVersionedSet(vs.Set().Union(previous.Set()), vs.APIVersion(), vs.Applied())
+		} else {
+			managed.Fields[manager] = vs
```

Comment:
I think it's idempotent? If you gave it 11 entries, it would merge the first one with itself and result in 11 entries out, like you said. And then when you run it again, the same thing would happen (the ancient changes entry would be the one which got merged with itself this time since it is always sorted first), and the result would be the same. I can make a test case for it though

Reply:
I see, it is sorted by age and the added one is always oldest (is it?).

So it is probably idempotent (yes please test) but I think it would be even better if "cap at 10" actually resulted in always outputting no more than 10 items (if possible), and resulted in the same number no matter whether an aggregate entry was present or not. So I guess I'm actually claiming this is off-by-one :)

Reply:
Off by the number of versions, yes. I think it could be edited to always give `entries <= max(apiVersions, 10)`

Reply:
would you expect the cap to be on appliers too? ex. If an object had three appliers, would you expect the number of updaters to be <= 7?

Reply:
It seems unlikely, but then what if you have more than 10 appliers? And what if you have 8 appliers, does it mean you have at most 2 non-appliers?

Reply:
> but then what if you have more than 10 appliers?

 Probably the same thing as it would do if you had more than 10 apiversions, just keep the number of updaters to a minimum


> And what if you have 8 appliers, does it mean you have at most 2 non-appliers?

 Yes